packages/terminal/src/browser/terminal-widget-impl.ts

```diff
@@ -404,4 +410,35 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         const { cols, rows } = this.term;
         this.shellTerminalServer.resize(this.terminalId, cols, rows);
     }
+
+    protected get enableCopy(): boolean {
+        return this.preferences['terminal.enableCopy'];
+    }
+
+    protected get enablePaste(): boolean {
+        return this.preferences['terminal.enablePaste'];
+    }
+
+    protected customKeyHandler(event: KeyboardEvent): boolean {
+        const keyBindings = KeyCode.createKeyCode(event).toString();
+        const ctrlCmdCopy = (isOSX && keyBindings === 'meta+c') || (!isOSX && keyBindings === 'ctrl+c');
+        const ctrlCmdPaste = (isOSX && keyBindings === 'meta+v') || (!isOSX && keyBindings === 'ctrl+v');
+        if (ctrlCmdCopy && this.enableCopy && this.term.hasSelection()) {
+            this.term.clearSelection();
+            return false;
+        }
+        if (ctrlCmdPaste && this.enablePaste) {
+            return false;
+        }
+        return true;
+    }
+    protected attachCustomKeyEventHandler(): void {
+        this.term.attachCustomKeyEventHandler(e => this.customKeyHandler(e));
+    }
+    protected async onPreferenceChanged(event: PreferenceChangeEvent<TerminalConfiguration>): Promise<void> {
+        const { preferenceName } = event;
+        if (preferenceName === 'terminal.enableCopy' || preferenceName === 'terminal.enablePaste') {
+            this.attachCustomKeyEventHandler();
```

Comment:
Can you explain why it is necessary to watch for preference changes?  I think it should not be necessary, since we look the preference value on each key stroke.  Also, does calling `attachCustomKeyEventHandler` multiple times adds a new handler to the list?  This would mean that each time the preferences change, the list of callback in xterm.js will grow.

Reply:
And regarding `attachCustomKeyEventHandler` xterm only ever attaches one. If a new declaration is called, it overwrites the previous.